loopy/kernel/dependency.py

```diff
@@ -0,0 +1,108 @@
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import islpy as isl
+from islpy import dim_type
+
+from loopy import LoopKernel
+from loopy.kernel.instruction import HappensAfter
+from loopy.translation_unit import for_each_kernel
+
+
+@for_each_kernel
+def add_lexicographic_happens_after(knl: LoopKernel) -> LoopKernel:
+    """Construct a sequential dependency specification between each instruction
+    and the instruction immediately before it. This dependency information
+    contains a lexicographic map which acts as a description of the precise,
+    statement-instance level dependencies between statements.
+    """
+
+    new_insns = []
+    for iafter, insn_after in enumerate(knl.instructions):
+        if iafter == 0:
+            new_insns.append(insn_after)
+
+        else:
+            insn_before = knl.instructions[iafter-1]
+
+            domain_before = knl.get_inames_domain(insn_before.within_inames)
+            domain_after = knl.get_inames_domain(insn_after.within_inames)
+
+            shared_inames = insn_before.within_inames & insn_after.within_inames
+
+            happens_after = isl.Map.from_domain_and_range(
+                    domain_before,
+                    domain_after)
+
+            for idim in range(happens_after.dim(dim_type.out)):
+                happens_after = happens_after.set_dim_name(
+                        dim_type.out,
+                        idim,
+                        happens_after.get_dim_name(dim_type.out, idim) + "'")
+
+            shared_inames_order_before = [
+                    domain_before.get_dim_name(dim_type.out, idim)
+                    for idim in range(domain_before.dim(dim_type.out))
+                    if domain_before.get_dim_name(dim_type.out, idim)
+                    in shared_inames]
+
+            shared_inames_order_after = [
+                    domain_after.get_dim_name(dim_type.out, idim)
+                    for idim in range(domain_after.dim(dim_type.out))
+                    if domain_after.get_dim_name(dim_type.out, idim)
+                    in shared_inames]
+
+            assert shared_inames_order_after == shared_inames_order_before
+            shared_inames_order = shared_inames_order_after
```

Comment:
This snippet here would imply that the order of variables in a set is significant. This is not the case anywhere else in loopy, and I would prefer to not start having it be the case now. This behavior is also not documented.

Reply:
So... I tried, but I ran out of time. :slightly_smiling_face: There's a bit of algorithmic work to be done here, and some important algorithmic pieces are in #690. We can discuss in more detail during our meeting Monday.

Rather than hijack this PR for my big pile of unrelated changes, that's now #864, which this depends on. The relevant pieces from #690 are also in #864, so you'll be able to rely on them.